src/Shared/CoreCLRAssemblyLoader.cs

```diff
@@ -73,29 +73,18 @@ public Assembly LoadFromPath(string fullPath)
             }
         }
 
-        private Assembly TryGetWellKnownAssembly(AssemblyLoadContext context, AssemblyName assemblyName)
-        {
-            if (!_wellKnownAssemblyNames.Contains(assemblyName.Name))
-            {
-                return null;
-            }
-
-            assemblyName.Version = _currentAssemblyVersion;
-
-            return context.LoadFromAssemblyName(assemblyName);
-        }
-
         private Assembly TryResolveAssembly(AssemblyLoadContext context, AssemblyName assemblyName)
         {
             lock (_guard)
             {
-                Assembly assembly = TryGetWellKnownAssembly(context, assemblyName);
-
-                if (assembly != null)
+                if (_wellKnownAssemblyNames.Contains(assemblyName.Name))
                 {
-                    return assembly;
+                    // Ensure we are attempting to load a matching version
+                    // of the Microsoft.Build.* assembly.
+                    assemblyName.Version = _currentAssemblyVersion;
```

Comment:
Since we don't vary assembly version across builds/releases any more, I think we might have to do more here, like force the load to come from `folder-next-to-current-assembly` rather than doing the full lookup.

Reply:
So I should extract the candidate path searching logic and call it from `TryGetWellKnownAssembly` with the executing assembly's location? something like:

```csharp
        private Assembly TryGetWellKnownAssembly(AssemblyLoadContext context, AssemblyName assemblyName)
        {
            if (!_wellKnownAssemblyNames.Contains(assemblyName.Name))
            {
                return null;
            }

            // Ensure we are attempting to load a matching version
            // of the Microsoft.Build.* assembly.
            assemblyName.Version = _currentAssemblyVersion;

            var searchPaths = new[] { Assembly.GetExecutingAssembly().Location };
            return TryResolveAssemblyFromPaths(context, assemblyName, searchPaths);
        }
```

Reply:
That sounds pretty good to me.

~I remember some funkiness around getting path-to-current-assembly in .NET Core, but I bet it's fixed now, that was a pre-1.0 memory. It _might_ be better to use `BuildEnvironmentHelper.Instance.CurrentMSBuildExePath`, but that might also accidentally trigger other loads~ nope, looks like that was all about targeting <netstandard1.3.

@tmeschter, can you dredge up any memories that might be relevant to this loader change?

Reply:
I'm not sure how much of this code is necessary--I had to work around multiple bugs in the CoreCLR to get *anything* to work at the time, and I would assume they have been fixed now.

I can't remember now why I even needed special handling of the well-known assemblies. Since we're always using the default `AssemblyLoadContext` I would assume that all loads of `Microsoft.Build.dll`, etc, would resolve to the ones shipped with the product (regardless of version), so why would we ever need to handle those in the resolve assembly event handler? But probably I've just forgotten how `AssemblyLoadContext` works. :-)

The commit messages are your best change for understanding what I was thinking at the time.